torch/optim/adam.py

```diff
@@ -83,11 +83,7 @@ def _init_group(
                 state = self.state[p]
                 # Lazy state initialization
                 if len(state) == 0:
-                    state['step'] = (
-                        torch.zeros((1,), dtype=torch.float, device=p.device)
-                        if group['capturable'] or group['fused']
-                        else torch.tensor(0.)
-                    )
+                    state['step'] = torch.tensor(0.0, device=p.device)
```

Comment:
It turns out that we used to have this code (without the if else) but the if-else is necessary because we _don't_ want the step to be on the parameter's device when capturable/fused is False. The reason is because kernel launches are quite expensive (especially for XLA) and so we move the step computation off the device intentionally.

Sorry for suggesting the change 😬 and let's add it back. Could you also add a comment explaining why the if else must exist?

Reply:
errrr, I was wondering the same before got confused. reverted the change